### DISCUSSION TITLE
Some answers by someone switching from v1

### DISCUSSION THREAD

**Discussion by kevinRoomies** [elrumordelaluz/reactour#430] [Q&A]
I'm trying to make my tour go over a menu that is has an absolute positionning but it cannot find it, is this a normal behavior

**Comment by elrumordelaluz** [comment]
I converted in a discussion. This way you can share all the questions. Keep in mind that would be easier to anyone to help if you share a sandbox with the code you are working on.

Thanks!

**Comment by kevinRoomies** [comment]
Making a sandbox would take too much time

**Reply by elrumordelaluz** [reply]
at least a reduced reproduction of the parts that creates issues you need help.

**Comment by ArutanSipdrae** [comment]
It was another case of the step trying to go to a dom element that isn't in the dom yet and doesn't update after it appearing despite having mutationObservables

**Reply by elrumordelaluz** [reply]
`mutationObservables` is useful in that cases.

**Reply by kevinRoomies** [reply]
that's what I thought, but it doesn't work, I've tried on all my steps with dom elements being added or removed it never refresh with them, resize is what refreshes

**Reply by elrumordelaluz** [reply]
keep in mind that you have to `observe` a parent element; try to create a sandbox where we can play and investigate.

**Reply by miko007** [reply]
here is a minimal sandbox, showing the problem *(or at least the manifestation of it, i have in my codebase)*:

https://codesandbox.io/s/cocky-mayer-gp5opl?file=/src/App.js

`mutationObservable` seems not to work, when location of the router is manipulated within the `action` attribute of a `step`.

**Reply by elrumordelaluz** [reply]
hi @miko007 and thank you for sharing your use case!
I encourage you to not use the `action` method to fire stuff that imply a re-render and calculation of the Tour.
Take a look at [this approach](https://codesandbox.io/s/blissful-lumiere-fkozqi?file=/src/App.js), using [`prevButton` and `nextButton` props](https://github.com/elrumordelaluz/reactour/tree/master/packages/tour#nextbutton-props-btnfnprops--void).

**Reply by miko007** [reply]
hey @elrumordelaluz that seems like an elegant solution at first glance, but it also means, that keyboard and "clicking on dots" navigation also has to be reimplemented...

i think that wanting to "tour" stuff on several different routes and visiting them during said tour automatically is a very regular use case. the code inside the steps like in the initial example did however work flawlessly with version 1.8.x.

maybe it would be helpful to have something like `beforeStepAction` or such, but i am not shure on this, as i just used the library and did not look into its internals.

**Reply by elrumordelaluz** [reply]
Oh you are absolutely right!

There are [some](https://github.com/elrumordelaluz/reactour/issues/102) [discussions](https://github.com/elrumordelaluz/reactour/pull/127) regarding pre and post `action`. I didn't included some logic in the new major release because there were lots of use-cases unsolved, like async calls, cancelling, and so on; and the logic started to be complicated.
Still thinking that could fit in future releases, help super welcome!

**Reply by elrumordelaluz** [reply]
@miko007 please give an eye on [this approach](https://codesandbox.io/s/inspiring-gates-m3b3xo?file=/src/App.js:655-666) using custom/controlled [setCurrentStep](https://github.com/elrumordelaluz/reactour/tree/master/packages/tour#setcurrentstep-dispatchreactsetstateactionnumber) and [currentStep](https://github.com/elrumordelaluz/reactour/tree/master/packages/tour#currentstep-number). Note that have to upgrade to `@reactour/tour@2.10.2` due to a bug fixed on keyboard navigation.

**Reply by miko007** [reply]
@elrumordelaluz well this is most likely the most elegant solution. i did not see that this option was available and i do not think anything more is needed in the api.

i would add that there is no need for `await` here, as the `navigate()` function returned by `useNavigate()` does not return a promise.

thank you very much for your support. maybe it should be considered to include this example in the examples section, as i guess more people will find it useful.

**Reply by elrumordelaluz** [reply]
sure! mind fixing the demo and adding to the readme?

**Reply by miko007** [reply]
@elrumordelaluz no problemo.

i fully cleaned up the example and provided PR #460
